### ISSUE TITLE
Poor albedo quality and visibility artifacts on TensoIR hotdog scene

### ISSUE THREAD

**Issue by SinonShizuku** [qbhan/RadioGS-public#2]
Thank you for your great work!

When reproducing your work on the **hotdog** scene of the **TensoIR** dataset, I obtained the following metrics for novel view synthesis:

## Novel View Synthesis

| Metric       | Value   |
|--------------|---------|
| PSNR         | 35.2822 |
| SSIM         | 0.9750  |
| LPIPS        | 0.0383  |
| SH PSNR      | 37.6411 |

## Material

| Metric                | Value   |
|-----------------------|---------|
| Albedo linear PSNR    | 23.8306 |
| Albedo linear SSIM    | 0.9164  |
| Albedo linear LPIPS   | 0.1116  |
| Albedo PSNR           | 25.7960 |
| Albedo SSIM           | 0.9439  |
| Albedo LPIPS          | 0.0883  |
| Normal MAE            | 4.3335  |

The albedo material quality seems not very good. Below is the visualization output from the inverse rendering training stage at the 19,000th iteration. The rendered terms such as visibility and indirect lighting all exhibit spotty dark artifacts:

<img width="1066" height="1600" alt="Image" src="https://github.com/user-attachments/assets/4aea27e7-5daa-4b2b-9132-ce5f15684b41" />

To investigate, I rendered the visibility using the following code (based on RadioGS):

```python
import os
import sys
import math
from argparse import ArgumentParser

import torch
from torchvision.utils import save_image

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from arguments import ModelParams, PipelineParams, get_combined_args
from gaussian_renderer.radiogs import sample_incident_rays
from diff_surfel_rasterization import GaussianRasterizationSettings, GaussianRasterizer
from scene import RadioGSModel, Scene
from utils.general_utils import safe_state


def rasterize_visibility(viewpoint_camera, gaussians, pipe, visibility):
    tanfovx = math.tan(viewpoint_camera.FoVx * 0.5)
    tanfovy = math.tan(viewpoint_camera.FoVy * 0.5)
    bg_color = torch.zeros(3, dtype=torch.float32, device="cuda")

    raster_settings = GaussianRasterizationSettings(
        image_height=int(viewpoint_camera.image_height),
        image_width=int(viewpoint_camera.image_width),
        tanfovx=tanfovx,
        tanfovy=tanfovy,
        bg=bg_color,
        scale_modifier=1.0,
        viewmatrix=viewpoint_camera.world_view_transform,
        projmatrix=viewpoint_camera.full_proj_transform,
        sh_degree=gaussians.active_sh_degree,
        campos=viewpoint_camera.camera_center,
        prefiltered=False,
        debug=pipe.debug,
    )
    rasterizer = GaussianRasterizer(raster_settings=raster_settings)
    means2D = torch.zeros_like(gaussians.get_xyz, dtype=gaussians.get_xyz.dtype, device="cuda")

    _, _, rendered_features, _, _ = rasterizer(
        means3D=gaussians.get_xyz,
        means2D=means2D,
        shs=None,
        colors_precomp=torch.zeros_like(gaussians.get_xyz),
        features=visibility,
        opacities=gaussians.get_opacity,
        scales=gaussians.get_scaling,
        rotations=gaussians.get_rotation,
        cov3D_precomp=None,
    )
    return rendered_features[:1]


def main():
    parser = ArgumentParser(description="Render one RadioGS visibility image")
    model = ModelParams(parser, sentinel=True)
    pipeline = PipelineParams(parser)
    parser.add_argument("--iteration", default=-1, type=int)
    parser.add_argument("--view_index", default=0, type=int)
    parser.add_argument("--output", required=True, type=str)
    parser.add_argument("--quiet", action="store_true")
    args = get_combined_args(parser)

    safe_state(args.quiet)

    with torch.no_grad():
        transmittance_min = getattr(args, "transmittance_min", 0.03)
        gaussians = RadioGSModel(args.sh_degree, transmittance_min=transmittance_min)
        scene = Scene(model.extract(args), gaussians, load_iteration=args.iteration, shuffle=False)
        gaussians.build_bvh()

        if gaussians.env_map is not None:
            gaussians.env_map.update_pdf()
        if scene.light_rotate and gaussians.env_map is not None:
            transform = torch.tensor(
                [[0, -1, 0], [0, 0, 1], [-1, 0, 0]],
                dtype=torch.float32,
                device="cuda",
            )
            gaussians.env_map.set_transform(transform)

        views = scene.getTestCameras()
        view = views[args.view_index % len(views)]

        dir_pp = gaussians.get_xyz - view.camera_center
        dir_pp_normalized = dir_pp / (torch.norm(dir_pp, dim=-1, keepdim=True) + 1e-6)
        normal_dummy = gaussians.get_normal(
            scaling_modifier=1.0,
            dir_pp_normalized=dir_pp_normalized,
        )
        incident_directions, incident_areas = sample_incident_rays(
            normal_dummy,
            is_training=False,
            sample_num=args.diffuse_sample_num,
        )
        gaussians.update_incidents_directions(incident_directions, incident_areas)
        gaussians.precompute_incidents(
            light_t_min=args.light_t_min,
            only_vis=False,
            back_culling=args.back_culling,
        )

        per_gaussian_visibility = gaussians.get_incident_visibility.mean(dim=1).clamp(0.0, 1.0)
        visibility = rasterize_visibility(
            view,
            gaussians,
            pipeline.extract(args),
            per_gaussian_visibility,
        ).clamp(0.0, 1.0)
        visibility = visibility.repeat(3, 1, 1)

        os.makedirs(os.path.dirname(args.output), exist_ok=True)
        save_image(visibility, args.output)
        print(args.output)
        print(
            "visibility stats:",
            "min", float(visibility.min()),
            "max", float(visibility.max()),
            "mean", float(visibility.mean()),
        )


if __name__ == "__main__":
    main()
```
The output visibility map is attached as an image (the same as the visual outputs during training):

<img width="800" height="800" alt="Image" src="https://github.com/user-attachments/assets/b08343b3-66ca-4134-80a9-da04c0f92641" />

Since your repository does not provide the `submodules` directory, I used the `submodules` directory from the IRGS project instead. Could you please provide some useful information to help diagnose this issue?

Also, are the submodules in your project exactly the same as those in IRGS? If not, could you share the relevant code from your submodules?

Thank you very much for your help!

**Comment by meltlis** [comment]
Hi, I’m running into the same issue in TensoIR hotdog scene and noticed this issue was closed. Could you share how you resolved it? Thanks!

**Comment by qbhan** [comment]
I am sorry about the issue.
I just pushed the submodules that include the missing rasterizer and ray tracing.

We have made slight adjustments on treating the dual-sided Gaussians during rasterization and ray tracing.
I conjecture that this may have created issue for reproduction.

I hope the updated submodules help you reproduce our method.

If you have any questions regarding the reproduction, please let me know.
